lib/systems/inspect.nix

```
@@ -17,6 +17,11 @@ rec {
     isAarch32      = { cpu = { family = "arm"; bits = 32; }; };
     isAarch64      = { cpu = { family = "arm"; bits = 64; }; };
     isMips         = { cpu = { family = "mips"; }; };
+    isMips32       = { cpu = { family = "mips"; bits = 32; }; };
+    isMips64       = { cpu = { family = "mips"; bits = 64; }; };
+    isMips64n32    = { cpu = { family = "mips"; }; abi = { abi = "n32"; }; };
+    isMips64n64    =[{ cpu = { family = "mips"; }; abi = { abi = "64";  };  }
+                     { cpu = { family = "mips"; bits = 64; }; abi = { name = "musl";  };  } ];
```

```suggestion
                     { cpu = { family = "mips"; bits = 64; };  } ];
```
do we need the Musl part? I wouldn't think we would want it?

---

So, the values in that attrset are defined to be sum-of-products patterns.  In other words, a platform is considered to be "Mips64n64" if it matches *any* of the patterns in the two-element list spanning lines 23-24.

The pattern on line 23 matches mips64[el]-*-gnuabi64; the pattern on line 24 matches mips64[el]-*-muslabi64.

Yeah, it's super-confusing.  But we do in fact need both of those lines, otherwise stuff like "lib.optionals stdenv.hostPlatform.isMips64n64" won't trigger for pkgsStatic.  I needed this when I was including a statically-linked chrpath in the bootstrap tarball (which is no longer necessary), but it is still required for cross-compiling to mips64el with static linking (static linking uses musl since glibc seems to be trying to abandon support for it).

The stuff in lib/systems/ is impressively cool, but could use a whole lot more explanatory commenting.  I should probably submit a PR to add comments while it's still fresh in my mind (had to reverse-engineer a fair bit of it).

Also it is probably time for a lot of the large attrsets in lib/systems to be broken out into separate directories/nixfiles.  I get merge conflicts from files in this directory when rebasing quite often.

---

> isMips64n64    = { cpu = { family = "mips"; bits = 64; }; abi = { abi = "64";  }; };

I think that will work.  Testing now.

---

Please check out https://github.com/a-m-joseph/nixpkgs/tree/nix-mips64el-democase, which contains this PR, #161162, the modification to `isMips64n64` above, and an `abort` in case the predicates don't trigger as expected.

When evaluating

```
nix-build . -A pkgs.pkgsCross.mips64el-linux-gnuabi64.pkgsStatic.nix_2_4
```

you'll find that the `boost` expression's `isMips64n64` doesn't trigger.  Dumping `builtins.toJSON stdenv.hostPlatform.parsed` gives us this:

```json
{
  "_type": "system",
  "abi": {
    "_type": "abi",
    "name": "musl"
  },
  "cpu": {
    "_type": "cpu-type",
    "bits": 64,
    "family": "mips",
    "name": "mips64el",
    "significantByte": {
      "_type": "significant-byte",
      "name": "littleEndian"
    }
  },
  "kernel": {
    "_type": "kernel",
    "execFormat": {
      "_type": "exec-format",
      "name": "elf"
    },
    "families": {},
    "name": "linux"
  },
  "vendor": {
    "_type": "vendor",
    "name": "unknown"
  }
}
```

So unfortunately when you ask nixpkgs to build you a statically-linked mips64el binary, you get a `hostPlatform.abi` with only two fields, and the only one of those which can be usefully matched against is `abi="musl"'.

That's why I wrote the extra line in `lib/systems/inspect.nix` the way I did.  Thanks for prompting me to remember why so it is documented!

---

The solution is not change this, but changing how `pkgsStatic` works.

---

I have implemented this in the two patches pushed on top of this PR.  Could you please review them?

> The solution is not change this, but changing how pkgsStatic works.

Wow, I never would have guessed that would have been the source of this problem.  To prevent hard-to-trace failures like this in the future, I have added a *commented-out* type check on the output of the parsedPlatform-staticification function.  I will submit a separate PR to uncomment it after this one is merged (in case the check causes build failures on obscure platforms).  We should still beef up the type check for parsed platforms at some point in the future.

---

> > In fact, unless you want to support something like:

```
mips-*-(gnu|musl)abi64
```

> > without the the 64 in the arch,

> I seem to remember this coming up as well while building `glibc`.

Ah yes, I stumbled across it again.

For posterity, to reproduce this issue, grab https://github.com/a-m-joseph/nixpkgs/tree/mips64el-support-first-steps-glibc-needs-to-parse-without-64-in-arch and then:

```
nix-build . -A pkgs.pkgsCross.mips64el-linux-gnuabi64.pkgsStatic.nix_2_4
```

... the build of `/nix/store/c36jzwynkpm4bp2ji1p2nszh5lm0rkp4-glibc-static-mips64el-unknown-linux-muslabi64-2.33-108.drv` will give you:

```
This configuration not matched in ../abi-tags
```

I really think we need to stop treating `musl` as an abi.  It is a libc.  There are platforms where libc and abi can be chosen independently.

---

> For posterity, to reproduce this issue, grab https://github.com/a-m-joseph/nixpkgs/tree/mips64el-support-first-steps-glibc-needs-to-parse-without-64-in-arch and then:

I have pushed an additional commit with a workaround for this problem.